### PR TITLE
fix(sql): ensure that set operations produce a single relation

### LINKED ISSUE #5878: bug: error on union of two parts of a pivoted table
### What happened?

Observed an error running this code, given the supplied data as `humanities.csv` file.
Applying `limit(n)` to the result got rid of the error.
I tried reproducing it with simpler data, but it observed no errors, so using the original.

# Code

```python
import ibis
from ibis.expr import selectors as sel
from ibis import _

from ibis.expr.types import Table

conn = ibis.duckdb.connect()
t = conn.read_csv("humanities.csv")


def compute_degrees_diff(tp: Table) -> Table:
    """Compute difference between the starting (earliest) number and the end one (latest)."""

    years_window = ibis.window(
        order_by=_["years"],
        group_by=_["Field of study"],
    )

    return (
        tp
        .mutate([
            _["degrees"].first().over(years_window).name("earliest_degrees"),
            _["degrees"].last().over(years_window).name("latest_degrees"),
        ])
        .mutate(diff=_["latest_degrees"] - _["earliest_degrees"])
    )


def top_and_bottom_movers():
    """Return a combined table with top and bottom ten items,
    based on the difference between start and end values.
    """

    tp = t.pivot_longer(
        cols=sel.matches(r"\d{4}-\d{2}"),
        names_to="years",
        values_to="degrees",
    )

    diff_agg = (
        compute_degrees_diff(tp)
        .agg(_["diff"].arbitrary().name("diff"), by="Field of study")
        .select(["Field of study", "diff"])
    )

    top_ten = diff_agg.order_by(_["diff"].desc()).limit(10)
    bottom_ten = diff_agg.filter(_["diff"] < 0).order_by(_["diff"]).limit(10)

    return top_ten.union(bottom_ten)


if __name__ == '__main__':
    result = top_and_bottom_movers()
    print(result.execute())  # ERROR!
    print(result.limit(100).execute())  # works OK with 'limit' clause

```

# Data
`humanities.csv`

```text
Field of study,1970-71,1975-76,1980-81,1985-86,1990-91,1995-96,2000-01,2005-06,2010-11,2011-12,2012-13,2013-14,2014-15,2015-16,2016-17,2017-18,2018-19,2019-20
Agriculture and natural resources,12674,19402,21886,17191,13363,21757,23766,23497,29203,31629,34304,35953,37028,37827,38782,40334,41373,41848
Architecture and related services,5570,9146,9455,9119,9781,8352,8480,9515,9831,9727,9757,9149,9090,8825,8579,8464,8806,9045
"Area, ethnic, cultural, gender, and group studies",2579,3577,2887,3021,4776,5633,6160,7878,8955,9228,8850,8275,7783,7840,7720,7717,7733,7767
Biological and biomedical sciences,35705,54154,43078,38395,39482,61014,60576,70602,89984,95850,100397,104657,109904,113794,116768,118661,121233,126590
Business,115396,143171,200521,236700,249165,226623,263515,318043,365133,367235,360887,358132,363741,371690,381109,386240,390490,387851
"Communication, journalism, and related programs ",10324,20045,29428,41666,51650,47320,58013,73658,83231,83771,84818,87612,90658,92551,93794,92289,92528,91752
Communications technologies ,478,1237,1854,1479,1397,853,1178,2987,4858,4983,4987,4991,5135,4824,4615,4231,4444,4424
Computer and information sciences and support    services,2388,5652,15121,42337,25159,24506,44142,47702,43066,47406,50961,55271,59586,64402,71416,79597,88638,97047
Education,176307,154437,108074,87147,110807,105384,105458,107235,104008,105656,104698,98838,91596,87221,85130,82599,83952,85057
Engineering,45034,38733,63642,77391,62448,62168,58209,66841,76356,81371,85987,92169,97852,106789,115671,121953,126692,128332
Engineering technologies,5148,7943,11713,19731,17303,15829,14660,14565,16741,17283,17010,16807,17253,17159,18119,18728,19620,19788
English language and literature/letters,63914,41452,31922,34083,51064,49928,50569,55094,52754,53765,52401,50464,45851,42797,41314,40002,39344,38036
Family and consumer sciences/human sciences,11167,17409,18370,13847,13920,14353,16421,20775,22438,23441,23930,24689,24584,25389,25080,24357,23956,23245
"Foreign languages, literatures, and linguistics",20988,17068,11638,11550,13937,14832,16128,19393,21705,21756,21647,20332,19493,18436,17643,16957,16605,16305
Health professions and related programs ,25221,53885,63665,64941,59636,85755,75537,91528,142890,163018,180437,197949,215478,228075,236931,243876,250650,257282
"Homeland security, law enforcement, and   firefighting ",2045,12507,13707,12704,16806,24810,25211,35319,47600,54091,60264,62416,62723,61159,59553,58121,57333,57044
Legal professions and studies,545,531,776,1223,1827,2123,1991,3302,4429,4595,4425,4513,4420,4243,4272,4242,4235,4262
"Liberal arts and sciences, general studies,   and humanities ",7481,18855,21643,21336,30526,33997,37962,44898,46717,46961,46790,45281,43649,43669,44103,44262,44222,42903
Library science ,1013,843,375,155,90,58,52,76,96,95,102,127,99,85,99,81,99,118
Mathematics and statistics ,24801,15984,11078,16122,14393,12713,11171,14760,17182,18841,20449,20987,21854,22778,24075,25253,26155,27216
Military technologies and applied sciences ,357,952,42,255,183,7,21,33,64,86,105,185,276,358,469,655,767,1156
Multi/interdisciplinary studies ,6324,13709,12986,13754,17774,26885,26478,30583,42473,45717,47658,48392,47556,48833,49631,51982,53303,55022
"Parks, recreation, leisure, fitness, and    kinesiology",1621,5182,5729,4623,4315,12974,17948,25489,35934,38998,42628,46047,49008,50912,53292,53887,53540,53749
Philosophy and religious studies ,8149,8447,6776,6396,7423,7900,9442,12841,14336,14104,14338,13776,12925,12133,11740,11872,11981,11889
Physical sciences and science technologies,21410,21458,23936,21711,16334,19716,18025,20521,24705,26664,28053,29307,30042,30483,31272,31542,31155,30738
Precision production ,0,0,0,2,2,12,31,55,43,37,36,37,48,51,32,45,47,39
Psychology,38187,50278,41068,40628,58655,73416,73645,88132,100906,109099,114446,117312,117573,117447,116859,116436,116550,119968
Public administration and social services ,5466,15440,16707,11887,14350,19849,19447,21986,26799,29695,31950,33483,34364,34433,35461,35637,35104,34356
Social sciences and history ,155324,126396,100513,93840,125107,126479,128036,161468,177169,178534,177767,173132,166971,161211,159097,159962,160678,161164
Theology and religious vocations ,3720,5490,5808,5510,4799,4933,6220,7687,7567,7845,7839,7865,7859,7826,7489,7252,7061,6855
Transportation and materials moving,0,225,263,1838,2622,3561,3748,5349,4941,4876,4661,4588,4730,4531,4708,4924,5051,5251
Visual and performing arts,30394,42138,40479,37241,42186,49296,61148,83292,93939,95806,97799,97414,95840,92979,91291,88507,89741,92332
Other and not classified,0,0,0,0,13258,1756,783,0,0,0,0,0,0,0,0,0,0,0
```

### What version of ibis are you using?

5.0.0

### What backend(s) are you using, if any?

DuckDB

### Relevant log output

```sh
Traceback (most recent call last):
  File "/Users/redacted/PycharmProjects/ibis-demo/bugs/union_alias.py", line 54, in <module>
    print(result.execute())  # ERROR!
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/ibis/expr/types/core.py", line 303, in execute
    return self._find_backend(use_default=True).execute(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/ibis/backends/base/sql/__init__.py", line 240, in execute
    self._log(sql)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/ibis/backends/base/sql/alchemy/__init__.py", line 408, in _log
    query_str = str(sql)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 294, in __str__
    return str(self.compile())
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 283, in compile
    return self._compiler(dialect, **kw)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 291, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 1269, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 710, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 755, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 2766, in visit_compound_select
    text = (" " + keyword + " ").join(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 2768, in <genexpr>
    c._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4551, in visit_select
    text = self._compose_select_body(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4699, in _compose_select_body
    [
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4700, in <listcomp>
    f._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 3893, in visit_cte
    inner = cte.element._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4551, in visit_select
    text = self._compose_select_body(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4699, in _compose_select_body
    [
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4700, in <listcomp>
    f._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4036, in visit_subquery
    return self.visit_alias(subquery, **kw)
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 3995, in visit_alias
    inner = alias.element._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4551, in visit_select
    text = self._compose_select_body(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4699, in _compose_select_body
    [
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 4700, in <listcomp>
    f._compiler_dispatch(
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/Users/redacted/.pyenv/versions/ibis-experiments/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 3817, in visit_cte
    raise exc.CompileError(
sqlalchemy.exc.CompileError: Multiple, unrelated CTEs found with the same name: 't3'

Process finished with exit code 1
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Fixes #5878.

~This PR ensures that we never generate an overlapping table name, even in the
case of a parent reference.~

~We do this by ticking the counter up by one every time an alias is generated,
no matter whether it's a parent reference or not.~

This PR ensures that set operations always have projections immediately after them.

### CODE REVIEW

File: ibis/tests/sql/snapshots/test_sqlalchemy/test_no_cross_join/out.sql [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+  t3.family,
+  t4.taken,
+  t4.person,
+  t4.quant,
+  t4.reading,
   t2.id AS id_y,
   t2.site,
   t2.dated
-FROM person AS t0
+FROM person AS t3, survey AS t4, person AS t0
```
Comment (cpcloud) [inline ibis/tests/sql/snapshots/test_sqlalchemy/test_no_cross_join/out.sql:12]: Looks like this is still a problem. Back to the drawing board. Serious deja vu right now.

Comment (cpcloud) [reply ibis/tests/sql/snapshots/test_sqlalchemy/test_no_cross_join/out.sql]: The code for this is left unchanged.

**Comment by cpcloud** [comment]
I'll add a test for the original bug.

File: ibis/expr/types/relations.py [hunk 2]
```diff
@@ -829,9 +829,12 @@ def difference(self, *tables: Table, distinct: bool = True) -> Table:
         >>> t1.difference().equals(t1)
         True
         """
-        return functools.reduce(
+        t = functools.reduce(
             functools.partial(ops.Difference, distinct=distinct), tables, self.op()
         ).to_expr()
+        if t.equals(self):
+            return t
+        return t.select(self.columns)
```
Comment (kszucs) [inline ibis/expr/types/relations.py:837]: So we effectively remove the table type from field selections?

Comment (cpcloud) [reply ibis/expr/types/relations.py:837]: Not sure I follow what you're asking. The difference between this PR and upstream is that in this PR we always project the results of a set operation. The union is still underneath the projection.

Comment (kszucs) [reply ibis/expr/types/relations.py:837]: Ideally `t.select()` should be able to optimize this out in the future as a no-op. 

It feels like that we have more serious problems in the analysis code actually causing this bug.

File: ibis/expr/types/relations.py [hunk 3]
```diff
@@ -829,9 +829,12 @@ def difference(self, *tables: Table, distinct: bool = True) -> Table:
         >>> t1.difference().equals(t1)
         True
         """
-        return functools.reduce(
+        t = functools.reduce(
             functools.partial(ops.Difference, distinct=distinct), tables, self.op()
         ).to_expr()
+        if t.equals(self):
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/types/relations.py:835]: Yep.

File: ibis/expr/types/relations.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:837]: Agree, we should be able to handle this elsewhere, but for now this fixes the issue without too much hassle.

Comment (cpcloud) [reply ibis/expr/types/relations.py:837]: Created #5917 for this.

**Review by jcrist — APPROVED** [review]
LGTM!

**Pull request merged**